Modules/mathmodule.c

```diff
@@ -2427,31 +2427,49 @@ vector_norm(Py_ssize_t n, double *vec, double max, int found_nan)
 /*[clinic input]
 math.dist
 
-    p: object(subclass_of='&PyTuple_Type')
-    q: object(subclass_of='&PyTuple_Type')
+    p: object
+    q: object
     /
 
 Return the Euclidean distance between two points p and q.
 
-The points should be specified as tuples of coordinates.
-Both tuples must be the same size.
+The points should be specified as sequences (or iterables) of
+coordinates.  Both inputs must have the same dimension.
 
 Roughly equivalent to:
     sqrt(sum((px - qx) ** 2.0 for px, qx in zip(p, q)))
 [clinic start generated code]*/
 
 static PyObject *
 math_dist_impl(PyObject *module, PyObject *p, PyObject *q)
-/*[clinic end generated code: output=56bd9538d06bbcfe input=937122eaa5f19272]*/
+/*[clinic end generated code: output=56bd9538d06bbcfe input=74e85e1b6092e68e]*/
 {
     PyObject *item;
     double max = 0.0;
     double x, px, qx, result;
     Py_ssize_t i, m, n;
-    int found_nan = 0;
+    int found_nan = 0, p_allocated = 0, q_allocated = 0;
     double diffs_on_stack[NUM_STACK_ELEMS];
     double *diffs = diffs_on_stack;
 
+    if (!PyTuple_Check(p)) {
```

Comment:
Why not call just `PySequence_Tuple()` (which is a no-op for tuples)? This would simplify the setup and the cleanup code.

It may be also worth to use `PySequence_Fast()` and `PySequence_Fast_GET_ITEM()` in the loop. This will save you allocating a new tuple. For general iterable `PySequence_Tuple()` creates a list and a tuple, but `PySequence_Fast()` creates just a list.

But you will need to call `PySequence_Fast_GET_SIZE()` in a loop because list's size can be changed when you convert items to the C double.

Reply:
I didn't want any external function calls on the fast path for the common case.